src/poliastro/twobody/propagation.py

```diff
@@ -88,24 +88,23 @@ def cowell(k, r, v, tofs, rtol=1e-11, *, events=None, ad=None, **ad_kwargs):
     tofs = tofs.to(u.s).value
 
     u0 = np.array([x, y, z, vx, vy, vz])
+    t0 = 0
 
     # Set the non Keplerian acceleration
     if ad is None:
 
         def ad(t0, u_, k_):
             return 0, 0, 0
 
-    f_with_ad = functools.partial(func_twobody, k=k, ad=ad, ad_kwargs=ad_kwargs)
+    p = [k, ad, ad_kwargs]
```

Comment:
All the test failures are because of this:

```
E           numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E           Internal error at <numba.core.typeinfer.CallConstraint object at 0x7f12ad537e80>.
E           Failed in nopython mode pipeline (step: analyzing bytecode)
E           CALL_FUNCTION_EX with **kwargs not supported
E           During: resolving callee type: type(CPUDispatcher(<function func_twobody at 0x7f12ce010af0>))
E           During: typing of call at /home/vsts/work/1/s/.tox/py38-fast/lib/python3.8/site-packages/nbkode/core.py (124)
E           
E           Enable logging at debug level for details.
E           
E           File ".tox/py38-fast/lib/python3.8/site-packages/nbkode/core.py", line 124:
E                       def _rhs(t, y):
E                           return rhs(t, y, params)
E                           ^
```

And I suspect it's because of the presence of `ad_kwargs` there, which is a dictionary.

Reply:
Yes , you are absolutely correct ... the `ad_kwargs` is incompatible with the function signature needed for correct execution of the function `nbkode.DOP853` ... I don't have the exact function call signature of the `ad` function so I will need to check the data type of `ad_kwargs` to confirm it being a dictionary...

Reply:
While looking in the documentation for `nbkode.DOP853` we can implement a nested list or nested array such as `p = [k, ad, [parameters for ad_kwargs]]` which can be resolved as `parameters` passed in the function is of `np.ndarray` the nature ... Yes, we can create a function for `ad` such that if the `ad_kwargs` doesnt have value we can pass `0` for the same & if it does have something we can instead pass those values but will make it a bit complicated & needs refactoring of `two_body` & tests for the same ...
A suggestion is, if we can execute the `ad` function outside of the execution it may be easier to understand & debug because I am doubting that `nbkode.DOP853` can pass a function as a parameter but need to check the same...

Reply:
Perhaps we should explore a whole different approach than the current one of passing `ad` and `ad_kwargs` to `func_twobody`. For example, we could change the signature of `cowell` to be something like this:

```
def cowell(k, r, v, tofs, rtol=1e-11, *, events=None, f=func_twobody):
```

(let's ignore the `events` for a moment)

In this scenario, the user could do something like:

```
custom_params = ...

def custom_f(t0, u_, k):  # Function signature expected by DOP853, only k allowed as extra parameter
    du_kep = func_twobody(t0, u_, k)  # Assume func_twobody knows nothing about perturbation accelerations

    # custom_ad and custom_params are defined in the outer scope
    # so custom_f can be a closure
    ax, ay, az = custom_ad(t0, u_, k, custom_params)
    du_ad = np.array([0, 0, 0, ax, ay, az])

    # Here we do the same that func_twobody does now with ad
    return du_kep + du_ad

...
cowell(k, r0, v0, tofs, f=custom_f)  # This should work!
```

Reply:
Copied this to #1052

Reply:
Yes looks promising ... One question ... Where is actually `ad` function defined ? ... Yes this is what I was proposing & can be done easily ... Will start on the same as directed :smile:

Reply:
`ad` can be any of the perturbation accelerations, check out https://docs.poliastro.space/en/latest/examples/Propagation%20using%20Cowells%20formulation.html and https://docs.poliastro.space/en/latest/examples/Natural%20and%20artificial%20perturbations.html

Reply:
This refactor was finally done in gh-1053.